stl/inc/ranges

```diff
@@ -22,15 +25,462 @@ _STL_DISABLE_CLANG_WARNINGS
+                    ; // nothing to do
+                }
+            }
+
+            return *this;
+        }
+
+        _Semiregular_box& operator=(_Semiregular_box&& _That) noexcept(is_nothrow_move_constructible_v<_Ty>) {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Engaged = false;
+            }
+
+            if (_That._Engaged) {
+                _Construct_in_place(_Val, _STD move(_That._Val));
+                _Engaged = true;
+            }
+
+            return *this;
+        }
+
+        constexpr explicit operator bool() const noexcept {
+            return _Engaged;
+        }
+
+        _NODISCARD constexpr _Ty& operator*() noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+        _NODISCARD constexpr const _Ty& operator*() const noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+
+    private:
+        union {
+            _Nontrivial_dummy_type _Dummy;
+            _Ty _Val;
+        };
+        bool _Engaged;
+    };
+
+    // clang-format off
+    template <_Copy_constructible_object _Ty>
+        requires default_initializable<_Ty>
+            && (copyable<_Ty>
+                || (is_nothrow_copy_constructible_v<_Ty>
+                    && (movable<_Ty> || is_nothrow_move_constructible_v<_Ty>)))
+    class _Semiregular_box<_Ty> {
+        // clang-format on
+    public:
+        _Semiregular_box() = default;
+
+        template <class... _Types>
+        constexpr _Semiregular_box(in_place_t, _Types&&... _Args) noexcept(
+            is_nothrow_constructible_v<_Ty, _Types...>) // strengthened
+            : _Val(_STD forward<_Types>(_Args)...) {}
+
+        _Semiregular_box(const _Semiregular_box&) = default;
+        _Semiregular_box(_Semiregular_box&&)      = default;
+        _Semiregular_box& operator=(const _Semiregular_box&) requires copyable<_Ty> = default;
+        _Semiregular_box& operator=(_Semiregular_box&&) requires movable<_Ty> = default;
+
+        _Semiregular_box& operator=(const _Semiregular_box& _That) noexcept {
+            _Val.~_Ty();
+            _Construct_in_place(_Val, _That._Val);
+            return *this;
+        }
+
+        _Semiregular_box& operator=(_Semiregular_box&& _That) noexcept {
+            _Val.~_Ty();
+            _Construct_in_place(_Val, _STD move(_That._Val));
+            return *this;
+        }
+
+        constexpr explicit operator bool() const noexcept {
+            return true;
+        }
+
+        _NODISCARD constexpr _Ty& operator*() noexcept {
+            return _Val;
+        }
+        _NODISCARD constexpr const _Ty& operator*() const noexcept {
+            return _Val;
+        }
+
+    private:
+        /* [[no_unique_address]] */ _Ty _Val = _Ty();
+    };
+#else // ^^^ no workaround / workaround vvv
+    template <class _Ty>
+    class _Semiregular_box_copy : public optional<_Ty> {
+    public:
+        using optional<_Ty>::optional;

```

Comment:
I know this is a temporary workaround, but should we have a `_Mybase` typedef? It's slightly strange to see `optional<_Ty>` repeatedly mentioned.